frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -1255,6 +1342,102 @@ pub mod pallet {
                 && latest_staked_value > Zero::zero()
         }
 
+        /// return the delegatee_address if exists to the max_depth
+        /// else return the delegator address if no delegatee exists
+        pub(crate) fn check_delegate(
+            delegator_address: T::AccountId,
+            contract_id: T::SmartContract,
+        ) -> Result<T::AccountId,DispatchError> {
+
+            // delegatee account must be set in order to claim rewards
+            ensure!(DelegateInfo::<T>::contains_key(&delegator_address, &contract_id), Error::<T>::DelegateeAccountNotSetForContract);
+
+            let mut previous_address: T::AccountId = delegator_address;
+
+            // return delegatee address following the mapping in DelegateInfo upto MAX_DEPTH_FOR_DELEGATION
+            for _ in 0..MAX_DEPTH_FOR_DELEGATION {
+                if let Ok(delegatee_address) =
+                    DelegateInfo::<T>::try_get(&previous_address, &contract_id)
+                {
+                    previous_address = delegatee_address;
+                } else {
+                    return Ok(previous_address);
+                }
+            }
```

Comment:
Interesting approach :slightly_smiling_face: 
I like that you limited the number of iterations.

However, solution doesn't have to be so complex.
It's enough to just allow this:
* **Alice** delegates rewards to **Bob**
* **Bob** delegates rewards to **Charlie**
* the second action just overwrites the storage item you've added